### PR TITLE
Feature/256 num stan threads

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

- Added logic to allow for specifying parallel chains and threads per chain
- Set os.environ['STAN_NUM_THREADS'] to match `sample` method args

Logic:  sample method takes arguments `chains`, `parallel_chains`, `threads_per_chain`.
~~A single chain will always run number of requested threads - allowing user to oversubscribe CPUs,
but if parallel_chains * threads_per_chains > CPUs available, number of parallel chains will
be decremented until this condition is met, until we're only running 1 chain at a time.~~

Logic as implemented:

- defaults:  num_chains = 4, parallel_chains = cpu_count(), threads_per_chain = 1
- specifying `chains` but not `parallel_chains` or `threads_per_chain` will run
`cpu_count()` chains in parallel single threaded
- specifying `threads_per_chain` but not `parallel_chains` will run `cpu_count()` chains
in parallel multi-threaded.
- specifying both does exactly what user requests.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
hi @ahartikainen - still updating docs and adding unit tests - but would be happy to get feedback.

**Comment by mitzimorris** [comment]
script to test from command line:
```
import os
from cmdstanpy.model import CmdStanModel

test_path = os.path.join('test', 'data')
logistic1_stan = os.path.join(test_path,'logistic1.stan')
redcard_data = os.path.join(test_path,'redcard.json')

logistic1 = CmdStanModel(stan_file=logistic1_stan, cpp_options={"STAN_THREADS":False})

fit = logistic1.sample(data=redcard_data, chains=3, parallel_chains=3,threads_per_chain=2)
```

File: cmdstanpy/model.py [hunk 1]
```diff
@@ -601,20 +606,56 @@ def sample(
...
-        if cores is None:
-            cores = max(min(cores_avail - 2, chains), 1)
-        if cores < 1:
+        if parallel_chains is None or parallel_chains > chains:
+            parallel_chains = max(min(cpu_count(), chains), 1)
+        if parallel_chains > cpu_count():
+            self._logger.warning(
+                'Requested %u parallel chains, but only %u CPUs,'
+                ' can only run %u chains in parallel.',
+                threads_per_chain, cpu_count(), cpu_count()
```
Comment (ahartikainen) [inline cmdstanpy/model.py:615]:
```suggestion
                parallel_chains, cpu_count(), cpu_count()
```

File: cmdstanpy/model.py [hunk 2]
```diff
@@ -601,20 +606,56 @@ def sample(
...
+            )
+            parallel_chains = max(min(cpu_count(), chains), 1)
+        if parallel_chains < 1:
```
Comment (ahartikainen) [inline cmdstanpy/model.py:618]: This is probably following CmdStanR logic. In R, it is common to set `-1` == max

But this is fine.

File: cmdstanpy/model.py [hunk 3]
```diff
@@ -601,20 +606,56 @@ def sample(
...
+        if threads_per_chain < 1:
             raise ValueError(
-                'Argument cores must be a positive integer value, '
-                'found {}.'.format(cores)
+                'Argument threads_per_chain must be a positive integer value, '
+                'found {}.'.format(threads_per_chain)
             )
-        if cores > cores_avail:
+        os.environ['STAN_NUM_THREADS'] = str(threads_per_chain)
+        if threads_per_chain > cpu_count():
```
Comment (ahartikainen) [inline cmdstanpy/model.py:631]: Should we lower the number to cpu_count or is there some benefit with a larger numbers?

File: cmdstanpy/model.py [hunk 4]
```diff
@@ -601,20 +606,56 @@ def sample(
...
+                'Requested threads_per_chain: %u > available CPUs: %u,'
+                ' will oversubcribe CPUs',
+                threads_per_chain,
+                cpu_count(),
             )
-            cores = cores_avail
+            if parallel_chains > 1:
+                parallel_chains = 1
+                self._logger.warning(", won't run chains in parallel")
+        if parallel_chains > 1:
```
Comment (ahartikainen) [inline cmdstanpy/model.py:641]: how much does this affect to runtime?

File: cmdstanpy/model.py [hunk 5]
```diff
@@ -1016,6 +1056,7 @@ def _run_cmdstan(
         """
         cmd = runset.cmds[idx]
         self._logger.info('start chain %u', idx + 1)
+        self._logger.info('threads: %s', os.environ['STAN_NUM_THREADS'])
```
Comment (ahartikainen) [inline cmdstanpy/model.py:1059]: STAN_NUM_THREADS could be None?

```suggestion
        self._logger.info('threads: %s', str(os.environ.get('STAN_NUM_THREADS')))
```

Comment (mitzimorris) [reply cmdstanpy/model.py:1059]: good catch - will amend logic - I don't think this needs to be reported at all, actually - will make this debug.

File: cmdstanpy/model.py (hunk 3 above)
Comment (mitzimorris) [reply cmdstanpy/model.py:631]: possibly some benefit - we do what the user asks - who are we to judge?

File: cmdstanpy/model.py (hunk 4 above)
Comment (mitzimorris) [reply cmdstanpy/model.py:641]: got rid of this code altogether - we don't try to override user requests - will run requested number of parallel chains with requested number of threads.

File: cmdstanpy/model.py (hunk 2 above)
Comment (mitzimorris) [reply cmdstanpy/model.py:618]: not sure I understand the comment about `-1 == max` ? as a test?  in any case, no longer doing this computation.

**Comment by ahartikainen** [comment]
Clever way to test threads

**Comment by mitzimorris** [comment]
hi @ahartikainen - this is ready for re-review - simplified logic for chains and parallel_chains.
lots of changes to the docs because when I started looking at the docstrings, I saw a lot of cruft that needed to be cleaned up.

also changed `stan_var_dims` to `stan_variable_dims` to bring it in line with CmdStanR.

I realize this is extra cleanup that perhaps should be done separately - mea cupla.

File: cmdstanpy/model.py [hunk 6]
```diff
@@ -752,13 +768,13 @@ def generate_quantities(
...
 
-        Output filenames are composed of the model name, a timestamp
-        in the form 'YYYYMMDDhhmm' and the chain id, plus the corresponding
-        filetype suffix, either '.csv' for the CmdStan output or '.txt' for
-        the console messages, e.g. `bernoulli_ppc-201912081451-1.csv`. Output
-        files  written to the temporary directory contain an additional
-        8-character random string, e.g.
-        `bernoulli_ppc-201912081451-1-5nm6as7u.csv`.
+        Output files are either written to a temporary directory or to the
+        specified output directory.  Ouput filenames correspond to the template
```
Comment (ahartikainen) [inline cmdstanpy/model.py:772]:
```suggestion
        specified output directory.  Output filenames correspond to the template
```

File: cmdstanpy/model.py [hunk 7]
```diff
@@ -905,12 +921,13 @@ def variational(
...
         or to a temporary directory which is deleted upon session exit.
 
-        Output filenames are composed of the model name, a timestamp
-        in the form 'YYYYMMDDhhmm' and the chain id, plus the corresponding
-        filetype suffix, either '.csv' for the CmdStan output or '.txt' for
-        the console messages, e.g. `bernoulli-201912081451-1.csv`. Output files
-        written to the temporary directory contain an additional 8-character
-        random string, e.g. `bernoulli-201912081451-1-5nm6as7u.csv`.
+        Output files are either written to a temporary directory or to the
+        specified output directory.  Ouput filenames correspond to the template
```
Comment (ahartikainen) [inline cmdstanpy/model.py:925]:
```suggestion
        specified output directory.  Output filenames correspond to the template
```

File: cmdstanpy/model.py [hunk 8]
```diff
@@ -1016,6 +1033,9 @@ def _run_cmdstan(
         """
         cmd = runset.cmds[idx]
         self._logger.info('start chain %u', idx + 1)
+        self._logger.debug(
+            'threads: %s', str(os.environ.get('STAN_NUM_THREADS'))
```
Comment (ahartikainen) [inline cmdstanpy/model.py:1037]: str is fine here, even when everything should be string in `os.environ`

File: cmdstanpy/stanfit.py [hunk 9]
```diff
@@ -610,17 +613,17 @@ def stan_variable(self, name: str) -> np.ndarray:
             ]
             var_dims = [dim0]
             var_dims.extend(dims)
-            return self.sample[
-                :, :, idxs[0] : idxs[len(idxs) - 1] + 1
-            ].reshape(tuple(var_dims), order='A')
+            return self.sample[:, :, idxs[0] : idxs[len(idxs) - 1] + 1].reshape(
```
Comment (ahartikainen) [inline cmdstanpy/stanfit.py:616]:
```suggestion
            return self.sample[:, :, idxs[0] : idxs[-1] + 1].reshape(
```

Last element of a sequence idx is -1

**Pull request merged**